autonomi/src/client/data_types/pointer.rs

```diff
@@ -71,60 +109,167 @@ impl Client {
             })?;
 
         // verify payment was successful
-        let (proof, _price) = match payment_proofs.get(&xor_name) {
-            Some((proof, price)) => (proof, price),
+        let (proof, price) = match payment_proofs.get(&xor_name) {
+            Some((proof, price)) => (Some(proof), price),
             None => {
-                error!("Pointer at address: {address:?} was already paid for");
-                return Err(PointerError::PointerAlreadyExists(address));
+                info!("Pointer at address: {address:?} was already paid for, update is free");
+                (None, &AttoTokens::zero())
+            }
+        };
+        let total_cost = *price;
+
+        let (record, payees) = if let Some(proof) = proof {
+            let payees = Some(proof.payees());
+            let record = Record {
+                key: NetworkAddress::from_pointer_address(address).to_record_key(),
+                value: try_serialize_record(
+                    &(proof, &pointer),
+                    RecordKind::DataWithPayment(DataTypes::Pointer),
+                )
+                .map_err(|_| PointerError::Serialization)?
+                .to_vec(),
+                publisher: None,
+                expires: None,
+            };
+            (record, payees)
+        } else {
+            let record = Record {
+                key: NetworkAddress::from_pointer_address(address).to_record_key(),
+                value: try_serialize_record(&pointer, RecordKind::DataOnly(DataTypes::Pointer))
+                    .map_err(|_| PointerError::Serialization)?
+                    .to_vec(),
+                publisher: None,
+                expires: None,
+            };
+            (record, None)
+        };
+
+        let get_cfg = GetRecordCfg {
+            get_quorum: Quorum::Majority,
+            retry_strategy: Some(RetryStrategy::default()),
+            target_record: None,
+            expected_holders: Default::default(),
+        };
+
+        let put_cfg = PutRecordCfg {
+            put_quorum: Quorum::All,
+            retry_strategy: None,
+            verification: Some((VerificationKind::Crdt, get_cfg)),
+            use_put_record_to: payees,
+        };
+
+        // store the pointer on the network
+        debug!("Storing pointer at address {address:?} to the network");
+        self.network
+            .put_record(record, &put_cfg)
+            .await
+            .inspect_err(|err| {
+                error!("Failed to put record - pointer {address:?} to the network: {err}")
+            })?;
+
+        Ok((total_cost, address))
+    }
+
+    /// Create a new pointer on the network
+    /// Make sure that the owner key is not already used for another pointer as each key is associated with one pointer
```

Comment:
This is regression from Registers where one key was able to be associated with many Registers, and makes Pointer significantly less useful than it could be. It at least adds complexity in forcing an application to manage multiple keys for different data, and store that information somewhere rather than be able to rely on a single default key and use it when updating a Pointer based on its address.

I hope that this limitation can be overcome though I'm not sure how difficult that would be!

Reply:
> This is regression from Registers where one key...

I think I see how this is not in fact a regression due to the properties of BLS keys. Can you confirm that:
- the app can generate the corresponding signing_key and owner from a main/default SecretKey and an `index: &[u8]`
- which means that the `index: &[u8]` can act as a proxy for the PointerAddress and the key needed to sign an update, providing you have the main/default SecretKey
- for Pointer access, only the PointerAddress is needed

This may be similar to how Registers worked but I think was hidden. Anyway, if I'm correct here just hit :+1: and I'll be :smile: